utils/wasm-builder/src/wasm_project.rs

```diff
@@ -220,15 +215,71 @@ impl WasmProject {
             let wasm_meta_path = self
                 .original_dir
                 .join([file_base_name, ".meta.txt"].concat());
-            let wasm_meta_hash_path = self.original_dir.join(".metahash");
 
             smart_fs::write_metadata(wasm_meta_path, metadata)
                 .context("unable to write `*.meta.txt`")?;
 
-            smart_fs::write(wasm_meta_hash_path, format!("{:?}", metadata.hash()))
-                .context("unable to write `.metahash`")?;
+            source_code = format!(
+                r#"{source_code}
+#[allow(improper_ctypes)]
+mod fake_gsys {{
+    pub type BlockNumber = u32;
+    pub type BufferStart = u8;
+    pub type Hash = [u8; 32];
+    pub type Length = u32;
+    pub type Value = u128;
+
+    #[repr(C, packed)]
+    #[derive(Default)]
+    pub struct LengthWithHash {{
+        pub length: Length,
+        pub hash: Hash,
+    }}
+
+    impl LengthWithHash {{
+        pub fn as_mut_ptr(&mut self) -> *mut Self {{
+            self as _
+        }}
+    }}
+
+    pub const PTR_SPECIAL: *const u128 = u32::MAX as *const u128;
+
+    extern "C" {{
+        pub fn gr_reply(
+            payload: *const BufferStart,
+            len: Length,
+            value: *const Value,
+            _delay: BlockNumber,
+            err_mid: *mut LengthWithHash,
+        );
+    }}
+}}
+
+use fake_gsys::*;
```

Comment:
I don't know whether it can yield some namespace conflicts with `orig_project::*`, but I'd use it explicitly: `fake_gsys::gr_reply`, `fake_gsys::LengthWithHash`, etc.